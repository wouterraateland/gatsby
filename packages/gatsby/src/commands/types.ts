import { PackageJson, Reporter } from "gatsby"

export interface ICert {
  keyPath: string
  certPath: string
  key: string
  cert: string
}

export interface IProgram {
  useYarn: boolean
  open: boolean
  openTracingConfigFile: string
  port: number
  host: string
  report: Reporter
  [`cert-file`]?: string
  [`key-file`]?: string
  directory: string
  https?: boolean
  sitePackageJson: PackageJson
  ssl?: ICert
}

// @deprecated
export enum BuildHTMLStage {
  DevelopHTML = `develop-html`,
  BuildHTML = `build-html`,
}

export enum Stage {
  Develop = `develop`,
  DevelopHTML = `develop-html`,
  BuildJavascript = `build-javascript`,
  BuildHTML = `build-html`,
}
