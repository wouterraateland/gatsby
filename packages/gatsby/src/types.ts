export interface IMatch {
  id: string
  context: {
    sourceMessage: string
    [key: string]: string
  }
  error?: Error | undefined
  [key: string]: unknown
}
