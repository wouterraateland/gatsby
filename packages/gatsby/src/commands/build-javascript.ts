import { Span } from "opentracing"
import webpack from "webpack"

import webpackConfig from "../utils/webpack.config"

import { IProgram } from "./types"

import { reportWebpackWarnings } from "../utils/webpack-error-utils"

export const buildProductionBundle = async (
  program: IProgram,
  parentSpan: Span
): Promise<webpack.Stats> => {
  const { directory } = program

  const compilerConfig = await webpackConfig(
    program,
    directory,
    `build-javascript`,
    null,
    { parentSpan }
  )

  return new Promise((resolve, reject) => {
    webpack(compilerConfig).run((err, stats) => {
      if (err) {
        return reject(err)
      }

      reportWebpackWarnings(stats)

      if (stats.hasErrors()) {
        return reject(stats.compilation.errors)
      }

      return resolve(stats)
    })
  })
}
