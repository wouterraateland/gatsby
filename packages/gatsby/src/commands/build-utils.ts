import fs from "fs-extra"
import path from "path"

import {
  remove as removePageHtmlFile,
  getPageHtmlFilePath,
} from "../utils/page-html"
import { remove as removePageDataFile, fixedPagePath } from "../utils/page-data"
import { IReduxState } from "../redux/types"

export const getChangedPageDataKeys = (
  state: IReduxState,
  cachedPageData: Map<string, string>
): string[] => {
  if (cachedPageData && state.pageData) {
    const pageKeys: string[] = []
    state.pageData.forEach((newPageDataHash: string, key: string) => {
      if (!cachedPageData.has(key)) {
        pageKeys.push(key)
      } else {
        const previousPageDataHash = cachedPageData.get(key)
        if (newPageDataHash !== previousPageDataHash) {
          pageKeys.push(key)
        }
      }
    })
    return pageKeys
  }

  return [...state.pages.keys()]
}

export const collectRemovedPageData = (
  state: IReduxState,
  cachedPageData: Map<string, string>
): string[] => {
  if (cachedPageData && state.pageData) {
    const deletedPageKeys: string[] = []
    cachedPageData.forEach((_value: string, key: string) => {
      if (!state.pageData.has(key)) {
        deletedPageKeys.push(key)
      }
    })
    return deletedPageKeys
  }
  return []
}

const checkAndRemoveEmptyDir = (publicDir: string, pagePath: string): void => {
  const pageHtmlDirectory = path.dirname(
    getPageHtmlFilePath(publicDir, pagePath)
  )
  const pageDataDirectory = path.join(
    publicDir,
    `page-data`,
    fixedPagePath(pagePath)
  )
  const hasFiles = fs.readdirSync(pageHtmlDirectory)

  // if page's html folder is empty also remove matching page-data folder
  if (!hasFiles.length) {
    fs.removeSync(pageHtmlDirectory)
    fs.removeSync(pageDataDirectory)
  }
}

const sortedPageKeysByNestedLevel = (pageKeys: string[]): string[] =>
  pageKeys.sort((a, b) => {
    const currentPagePathValue = a.split(`/`).length
    const previousPagePathValue = b.split(`/`).length
    return previousPagePathValue - currentPagePathValue
  })

export const removePageFiles = async (
  publicDir: string,
  pageKeys: string[]
): Promise<void> => {
  const removePages = pageKeys.map(pagePath =>
    removePageHtmlFile({ publicDir }, pagePath)
  )

  const removePageData = pageKeys.map(pagePath =>
    removePageDataFile({ publicDir }, pagePath)
  )

  return Promise.all([...removePages, ...removePageData]).then(() => {
    // Sort removed pageKeys by nested directories and remove if empty.
    sortedPageKeysByNestedLevel(pageKeys).forEach(pagePath => {
      checkAndRemoveEmptyDir(publicDir, pagePath)
    })
  })
}
