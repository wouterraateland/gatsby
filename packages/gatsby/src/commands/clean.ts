import fs from "fs-extra"
import path from "path"

import {
  userPassesFeedbackRequestHeuristic,
  showFeedbackRequest,
} from "../utils/feedback"
import { IProgram } from "./types"

module.exports = async function clean(program: IProgram): Promise<void> {
  const { directory, report } = program

  const directories = [`.cache`, `public`]

  report.info(`Deleting ${directories.join(`, `)}`)

  await Promise.all(
    directories.map(dir => fs.remove(path.join(directory, dir)))
  )

  report.info(`Successfully deleted directories`)

  if (await userPassesFeedbackRequestHeuristic()) {
    showFeedbackRequest()
  }
}
