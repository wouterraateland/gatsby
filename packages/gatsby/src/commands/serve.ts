import path from "path"
import openurl from "better-opn"
import fs from "fs-extra"
import compression from "compression"
import express from "express"
import chalk from "chalk"
import { match as reachMatch } from "@reach/router/lib/utils"
import onExit from "signal-exit"
import report from "gatsby-cli/lib/reporter"

import telemetry from "gatsby-telemetry"

import { detectPortInUseAndPrompt } from "../utils/detect-port-in-use-and-prompt"
import getConfigFile from "../bootstrap/get-config-file"
import preferDefault from "../bootstrap/prefer-default"
import { IProgram } from "./types"

interface IMatchPath {
  path: string
  matchPath: string
}

interface IServeProgram extends IProgram {
  prefixPaths: boolean
}

onExit(() => {
  telemetry.trackCli(`SERVE_STOP`)
})

const readMatchPaths = async (
  program: IServeProgram
): Promise<IMatchPath[]> => {
  const filePath = path.join(program.directory, `.cache`, `match-paths.json`)
  let rawJSON = `[]`
  try {
    rawJSON = await fs.readFile(filePath, `utf8`)
  } catch (error) {
    report.warn(error)
    report.warn(
      `Could not read ${chalk.bold(
        `match-paths.json`
      )} from the .cache directory`
    )
    report.warn(
      `Client-side routing will not work correctly. Maybe you need to re-run ${chalk.bold(
        `gatsby build`
      )}?`
    )
  }
  return JSON.parse(rawJSON) as IMatchPath[]
}

const matchPathRouter = (
  matchPaths: IMatchPath[],
  options: {
    root: string
  }
) => (
  req: express.Request,
  res: express.Response,
  next: express.NextFunction
): void => {
  const { url } = req
  if (req.accepts(`html`)) {
    const matchPath = matchPaths.find(
      ({ matchPath }) => reachMatch(matchPath, url) !== null
    )
    if (matchPath) {
      return res.sendFile(
        path.join(matchPath.path, `index.html`),
        options,
        err => {
          if (err) {
            next()
          }
        }
      )
    }
  }
  return next()
}

module.exports = async (program: IServeProgram): Promise<void> => {
  telemetry.trackCli(`SERVE_START`)
  telemetry.startBackgroundUpdate()
  let { prefixPaths, port, open, host } = program
  port = typeof port === `string` ? parseInt(port, 10) : port

  const { configModule } = await getConfigFile(
    program.directory,
    `gatsby-config`
  )
  const config = preferDefault(configModule)

  const { pathPrefix: configPathPrefix } = config || {}

  const pathPrefix = prefixPaths && configPathPrefix ? configPathPrefix : `/`

  const root = path.join(program.directory, `public`)

  const app = express()
  const router = express.Router()

  app.use(telemetry.expressMiddleware(`SERVE`))

  router.use(compression())
  router.use(express.static(`public`))
  const matchPaths = await readMatchPaths(program)
  router.use(matchPathRouter(matchPaths, { root }))
  router.use((req, res, next) => {
    if (req.accepts(`html`)) {
      return res.status(404).sendFile(`404.html`, { root })
    }
    return next()
  })
  app.use(function(
    _: express.Request,
    res: express.Response,
    next: express.NextFunction
  ) {
    res.header(`Access-Control-Allow-Origin`, `*`)
    res.header(
      `Access-Control-Allow-Headers`,
      `Origin, X-Requested-With, Content-Type, Accept`
    )
    next()
  })
  app.use(pathPrefix, router)

  const startListening = (): void => {
    app.listen(port, host, () => {
      const openUrlString = `http://${host}:${port}${pathPrefix}`
      report.info(`gatsby serve running at: ${chalk.bold(openUrlString)}`)
      if (open) {
        report.info(`Opening browser...`)
        Promise.resolve(openurl(openUrlString)).catch(() =>
          report.warn(`Browser not opened because no browser was found`)
        )
      }
    })
  }

  try {
    port = await detectPortInUseAndPrompt(port)
    startListening()
  } catch (e) {
    if (e.message === `USER_REJECTED`) {
      return
    }

    throw e
  }
}
