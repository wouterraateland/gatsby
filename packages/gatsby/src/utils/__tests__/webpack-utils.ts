import { createWebpackUtils } from "../webpack-utils"
import { Stage, IProgram } from "../../commands/types"

jest.mock(`../browserslist`, () => {
  return {
    getBrowsersList: (): string[] => [],
  }
})

jest.mock(`babel-preset-gatsby/package.json`, () => {
  return {
    version: `1.0.0`,
  }
})

let config

beforeAll(() => {
  config = createWebpackUtils(Stage.Develop, {} as IProgram)
})

describe(`webpack utils`, () => {
  describe(`js`, () => {
    it(`adds .js rule`, () => {
      expect(config.rules.js).toEqual(expect.any(Function))
    })

    it(`returns default values without any options`, () => {
      const rule = config.rules.js([])

      expect(rule).toMatchSnapshot({
        use: [
          {
            loader: expect.stringContaining(`babel-loader`),
          },
        ],
      })
    })
    describe(`include function`, () => {
      let js
      beforeAll(() => {
        js = config.rules.js({
          modulesThatUseGatsby: [
            {
              name: `gatsby-seo`,
              path: `/Users/sidharthachatterjee/Code/gatsby-seo-test/node_modules/gatsby-seo`,
            },
          ],
        })
      })

      it(`includes source files from user code`, () => {
        expect(
          js.include(
            `/Users/sidharthachatterjee/Code/gatsby-seo-test/src/pages/index.js`
          )
        ).toEqual(true)
      })
      it(`includes files from .cache`, () => {
        expect(
          js.include(
            `/Users/sidharthachatterjee/Code/gatsby-seo-test/.cache/production-app.js`
          )
        ).toEqual(true)
      })
      it(`includes dependencies that use gatsby`, () => {
        expect(
          js.include(
            `/Users/sidharthachatterjee/Code/gatsby-seo-test/node_modules/gatsby-seo/index.js`
          )
        ).toEqual(true)
      })
      it(`does not include other dependencies`, () => {
        expect(
          js.include(
            `/Users/sidharthachatterjee/Code/gatsby-seo-test/node_modules/react/index.js`
          )
        ).toEqual(false)
      })
      it(`includes gatsby-browser.js`, () => {
        expect(
          js.include(
            `/Users/sidharthachatterjee/Code/gatsby-seo-test/gatsby-browser.js`
          )
        ).toEqual(true)
      })
    })
  })
  describe(`dependencies`, () => {
    it(`adds dependency rule`, () => {
      expect(config.rules.dependencies).toEqual(expect.any(Function))
    })
    it(`returns default values without any options`, () => {
      const rule = config.rules.dependencies()

      expect(rule).toMatchSnapshot({
        use: [
          {
            loader: expect.stringContaining(`babel-loader`),
          },
        ],
      })
    })
    describe(`exclude function`, () => {
      let dependencies
      beforeAll(() => {
        dependencies = config.rules.dependencies({
          modulesThatUseGatsby: [
            {
              name: `gatsby-seo`,
              path: `/Users/sidharthachatterjee/Code/gatsby-seo-test/node_modules/gatsby-seo`,
            },
          ],
        })
      })
      it(`excludes source files from user code`, () => {
        expect(
          dependencies.exclude(
            `/Users/sidharthachatterjee/Code/gatsby-seo-test/src/pages/index.js`
          )
        ).toEqual(true)
      })
      it(`excludes files from .cache`, () => {
        expect(
          dependencies.exclude(
            `/Users/sidharthachatterjee/Code/gatsby-seo-test/.cache/production-app.js`
          )
        ).toEqual(true)
      })
      it(`excludes dependencies that use gatsby`, () => {
        expect(
          dependencies.exclude(
            `/Users/sidharthachatterjee/Code/gatsby-seo-test/node_modules/gatsby-seo/index.js`
          )
        ).toEqual(true)
      })
      it(`excludes babel-runtime`, () => {
        expect(
          dependencies.exclude(
            `/Users/misiek/test/pr15285/node_modules/@babel/runtime/helpers/interopRequireDefault.js`
          )
        ).toEqual(true)
      })
      it(`excludes core-js`, () => {
        expect(
          dependencies.exclude(
            `/Users/misiek/test/pr15285/node_modules/core-js/modules/es6.array.iterator.js`
          )
        ).toEqual(true)
      })
      it(`includes dependencies that don't use gatsby`, () => {
        expect(
          dependencies.exclude(
            `/Users/sidharthachatterjee/Code/gatsby-seo-test/node_modules/react/index.js`
          )
        ).toEqual(false)
      })
      it(`excludes gatsby-browser.js`, () => {
        expect(
          dependencies.exclude(
            `/Users/sidharthachatterjee/Code/gatsby-seo-test/gatsby-browser.js`
          )
        ).toEqual(true)
      })
    })
  })
})
