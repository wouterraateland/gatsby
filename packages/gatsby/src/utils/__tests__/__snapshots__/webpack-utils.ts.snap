// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`webpack utils dependencies returns default values without any options 1`] = `
Object {
  "exclude": [Function],
  "test": /\\\\\\.\\(js\\|mjs\\)\\$/,
  "type": "javascript/auto",
  "use": Array [
    Object {
      "loader": StringContaining "babel-loader",
      "options": Object {
        "babelrc": false,
        "cacheIdentifier": "develop---gatsby-dependencies@1.0.0",
        "compact": false,
        "configFile": false,
        "presets": Array [
          "<PROJECT_ROOT>/packages/babel-preset-gatsby/dependencies.js",
        ],
        "sourceMaps": false,
      },
    },
  ],
}
`;

exports[`webpack utils js returns default values without any options 1`] = `
Object {
  "include": [Function],
  "test": /\\\\\\.\\(js\\|mjs\\|jsx\\)\\$/,
  "type": "javascript/auto",
  "use": Array [
    Object {
      "loader": StringContaining "babel-loader",
      "options": Object {
        "compact": false,
        "configFile": true,
        "stage": "develop",
      },
    },
  ],
}
`;
